### DISCUSSION TITLE
Airflow Breeze on M1 fail to initialize with Mysql

### DISCUSSION THREAD

**Discussion by ibrahimdurmus** [apache/airflow#28429] [General] [answered]
I was trying to get breeze working on my M1 chip and followed the instruction in [here](https://github.com/apache/airflow/blob/main/BREEZE.rst#installation). First part went fine but when I tried to run airflow runtime
 
```breeze --python 3.7 --backend mysql --mysql-version 8```
 
I am getting the following error stating mysql is not supported on ARM. As far as I searched in discussion board, couldn't locate a remedy. Sounds like mysql image is not in docker registry for ARM?

``` MySQL is not supported on ARM architecture. ```
 
Any one seen something similar? The closest thread - not the same though - I was able to find is located here : https://github.com/apache/airflow/discussions/27217
 
Cc: @bisk if he has any insight.

thanks

**Comment by potiuk** [comment]
Correct. This is one of limitations of M1. I have not checked recently but Oracle did not release neither client for debian nor native server image working on ARM. Until they do - ARM support for myself does not work (same story with mssql)

**Comment by potiuk** [comment, marked as answer]
Ok. Update. Seems that 10 days ago mysql pushed MySQL 8 ARM image:


<img width="1332" alt="Screenshot 2022-12-17 at 09 51 07" src="https://user-images.githubusercontent.com/595491/208234091-5970d081-9a19-4251-8ebf-aa8318b609ad.png">

But still MySQL Client ARM is only supported for Oracle Linux/RedHat/Centos:

https://www.mysql.com/fr/support/supportedplatforms/database.html

<img width="1162" alt="Screenshot 2022-12-17 at 09 53 35" src="https://user-images.githubusercontent.com/595491/208234161-715e6817-53a5-4752-9312-2d43df97ccc2.png">

Can't do much on this one. We have a possibility of supporting Centos as base image:  https://github.com/apache/airflow/pull/24825 but this is something that is a bit lower priority.

Potentially we could consider enabling using emulated AMD images, but this is terrible. ~ 16 times slower because of emulation - basically unusable.

**Reply by Taragolis** [reply]
@potiuk that is returns me back to the early 2020, that was last time when I tried to run something related to MySQL on armv7/armv8 debian-based  linux. So... in 2020 Oracle already provided client binaries for RPM based linux and if you need it on debian you have to build from sources and it required something like: specific version of Boost, patch them and about 8 GB or RAM. Unfortunetly MySQL is not a PostgreSQL which literally could built on very weak hardware.

And yeah still have issues with MySQL DBA-API clients
- [mysql-connector-python](https://pypi.org/project/mysql-connector-python/#files), as you mention Oracle doesn't build for linux ARMv8/ARM64
- [mysqlclient](https://pypi.org/project/mysqlclient/#files) could be built but required mysql dev libraries.

**Reply by potiuk** [reply]
Yeah. I would not want to build "against" official support for MySQL. I do not want to go the rabbit hole - even if we manage to fix it, there is high probability it will hold us back in the future when for example we want to move away from bullseye to bookworm when it is ready. 

I prefer in this case to stay with "postgres-only" for ARM. Yes it makes people who have M1/ARM not able to contribute to mysql-specific code, but it also reflects our recommendations to use Postgres. This is only one of many problems that Airflow has with MySQL and for me it makes much more sense that M1 is "Postgres-only". We never promised anything else, M1 from the very beginning does not support MySQL nor MsSQL, and I think it might stay like that forever and officially if Oracle and Microsoft won't support our Debian distribution.

At the end - our image is a reference one and we can make decision which platforms are supported on ARM. If we start supporting MySQL and Microsoft with some "hacks" rather than official support from Oracle/Microsoft - we are shooting ourselves in the foot. It might hold us back in the future. 

We might add a few small things though - the MySQL tests that are failing now should be just skipped on ARM - so that a user can run all tests and not see faiiling tests. I will do that shortly.

**Reply by ibrahimdurmus** [reply]
thanks for inputs @potiuk @Taragolis , for now, i am ok to use sqlite (or postgres) to continue on what i was using.

regarding to ```one of many problems that Airflow has with MySQL```, not sure if there was ever a compiled list of reasons for that statement? either way i would be researching on it, since in my current setup, we have a heavy reliance on mysql.

**Reply by potiuk** [reply]
see for example this discussion @ibrahimdurmus https://lists.apache.org/thread/dp78j0ssyhx62008lbtblrc856nbmlfb

**Reply by Taragolis** [reply]
MySQL + Python + ARM it is general issue and also effected providers which use MySQL, but we unfortunetly can't do anything here.
It also could affect users who decide migrate their compute resources from x86_64 to ARM.
But but thankfully it not a general case right now.

> We never promised anything else, M1 from the very beginning does not support MySQL nor MsSQL

BTW, Oracle provide `mysql-connector-python` for macos on M1, but not for linux arm 🤦

**Reply by potiuk** [reply]
> BTW, Oracle provide `mysql-connector-python` for macos on M1, but not for linux arm 🤦

Yep. It will likely work on local venv, but not in Breeze.

**Reply by potiuk** [reply]
And as consequence - not in our prod Docker image.

**Reply by Taragolis** [reply]
Yep, I mean technically we could skip entire test module only if required modules not found so we could add into specific modules something like that. 

```python
import pytest

pytest.importorskip("MySQLdb", "Couldn't find `mysqlclient` package.")
pytest.importorskip("mysql", "Couldn't find `mysql-connector-python` package.")

# Actual imports below this point
import MySQLdb
import mysql.connector
```

**Reply by potiuk** [reply]
Yep. That's in my local changes  now ;). Soon to be submitted

**Reply by potiuk** [reply]
But I actually prefer try/except especially when there are few related imports.
